### ISSUE TITLE
Various interaction issues with 1.0.5 and the keyboard ... 

### ISSUE THREAD

**Issue by sjmcdowall** [AbdulRahmanAlHamali/flutter_typeahead#61]
<img width="472" alt="screen shot 2019-02-28 at 10 22 32 am" src="https://user-images.githubusercontent.com/2034925/53577468-b78f9a00-3b43-11e9-9a23-7cd8f309ea32.png">
Our users are trying our app and we have some pretty critical UX issues that are appearing that we need some help with quickly .. (we hope to launch next week).

@KaYBlitZ -- any chance you can look at these?   THey are all keyboard related.. and I think part of the solution IS going to use that 3rd party keyboard detection thing .. 

Problem 1 -- on iOS when we enter a screen with a Typeahead the keyboard isn't appearing (see screen shot on an XR) and there is no way to make it appear since the focus is already in the TextField (that anyone can figure out).  The keyboard DOES appear on Android .. %K on iOS causes the keyboard to appear but -- shouldn't it always appear on entry / focus ?? 

2) On Android if I click on the TextField the keyboard reappears .. this does not happen on iOS

3) If keyboard is open -- no way to scroll the list to select an entry hidden by keyboard .. 

4) If we implement the "close list on keyboard close" -- how would the user EVER scroll the list?  I thought we recalculated the "display" size to accomodate the keyboard being open so the list could scroll? 

5) Assuming we fix 4, then we need to implement the "on keyboard close" dismiss the suggestions box option.  We need this because there are some critical buttons being hidden by the suggestions box that allow the user to continue the process but they can't get access to them so we need a way to toggle showing the suggestions box or not .. and not many options on how to do it that I can see.

Phew.

/Steve

**Comment by KaYBlitZ** [comment]
I can take a look later in the day. Haven't encountered problems 1 & 2, but I will try to replicate it. Problem 3 shouldn't be an issue if the direction is down. The suggestions box should resize between the bottom of the textbox and just above the keyboard so the keyboard should never be hiding anything. Only problem would be if there isn't enough space between the textbox and keyboard so the options cannot be clearly seen. But that may just be a UI design issue, since the keyboard would be covering those options anyways. If the keyboard is hidden, then the list will take up the full space. If the direction is up, then the initial options can be covered  by the keyboard since the suggestions box start just above the textbox. We would have to calculate the offset and position accordingly or just have the user close the keyboard.

Problems 4 & 5 seems like a design issue we would have to decide on. We can't have both at the same time. We can have a property, maybe `hideSuggestionsOnKeyboardClose` to toggle between the two behaviors? For my app I was having the user's scroll through the list with the keyboard open. There was enough space for that.

Abdul is fine with using the 3rd party library. I can work on it later. What do you think about the property idea? We can have it default to true, since that seems to be the way it would generally be used. People can then set it false if they want the box still opened when they close the keyboard.

**Comment by sjmcdowall** [comment]
Let me think about it -- because it's like a catch-22 thing where (I know impossible) we may want both actions. :)  However, if we can get the scrolling to work properly with the keyboard showing on Android .. then having the suggestions box gone with keyboard no-show would work.  See screen shot on Android showing the list -- both with keyboard and without.  

When the keyboard is showing, it won't scroll for me.  Nothing I do gets it to scroll.  I KNOW this worked not long ago and it was working perfectly.  <sigh> .. 

I think it's more important to allow the suggestions Box to close if no keyboard since the user can control that if the scrolling would actually work.. So #5 is good but havnig an OPTION for it is even better! :)

<img width="583" alt="screen shot 2019-02-28 at 12 44 24 pm" src="https://user-images.githubusercontent.com/2034925/53586672-a9974480-3b56-11e9-8f3a-703de0e60e79.png">
<img width="583" alt="screen shot 2019-02-28 at 12 44 47 pm" src="https://user-images.githubusercontent.com/2034925/53586673-a9974480-3b56-11e9-9450-8a63510ed2fe.png">

**Comment by KaYBlitZ** [comment]
From your images, the suggestions box is rendering with its full height, which is why scrolling doesn't work. That shouldn't be happening with the latest version. I am using the example included with typeahead and it's working properly. I would double-check that you are using 1.0.5 and maybe do a rebuild? Can you post the code that you are using to create the TypeAheadField? It is possible that something might be conflicting with the resize function call....

For problem 1, I assume that you are using the iOS simulator. By any chance is the simulator keyboard disabled in the options? It might be toggled off.

**Comment by sjmcdowall** [comment]
Yeah problrem #1 could easily be because of the simulator -- I can toggle the keyboard with %K .. so that's really not a big deal (and maybe a red herring).  I am for sure using 1.0.5. -- here is the pubspec.lock .. 
```
 flutter_typeahead:
    dependency: "direct main"
    description:
      name: flutter_typeahead
      url: "https://pub.dartlang.org"
    source: hosted
    version: "1.0.5"
```
Here is the code I am using -- is it possible the fixes didn't get into TypeAheadFormField ?

```

    return TypeAheadFormField<dynamic>(
        getImmediateSuggestions: true,
        textFieldConfiguration: TextFieldConfiguration<String>(
          focusNode: _myFocusNode,
          // autofocus: true,
          autofocus: false,
          style: DefaultTextStyle.of(context)
              .style
              .copyWith(fontStyle: FontStyle.italic),
          decoration: InputDecoration(
            border: const OutlineInputBorder(),
            filled: true,
            hintText: searchPrompt,
          ),
        ),
        suggestionsCallback: (String pattern) async {
          logger.fine(
              '[cETS - TypeAheadField() - suggestionsCallback] - Fired! With pattern [$pattern]');

          // See if we have searched this thing already..
          if (pattern == _oldPattern) {
            logger.fine(
                '[cETS - TypeAheadField() - suggestionsCallback] - Pattern == oldPattern .. Using cached Data');
            return _templates;
          }

          // New search string..go get the results..
          _oldPattern = pattern;
          _templates = await _getSuggestions(pattern);

          return _templates;
        },
        itemBuilder: (BuildContext context, dynamic suggestion) {
          // logger.fine('[cETS - TypeAheadField() - itemBuilder] - Fired!');
          return ListTile(
            contentPadding:
                const EdgeInsets.symmetric(vertical: 0.0, horizontal: 15.0),
            title: Text(
              suggestion['name'],
              style: const TextStyle(fontSize: 12.0),
            ),
            dense: true,
            // subtitle: Text(suggestion['descr']),
          );
        },
        validator: (String value) {
          logger.fine('Validator called value is $value');
          if (globals.user.currentOrg.isRestrictedSender &&
              widget.eventData.eventCategory == null) {
            return 'Please select a template';
          }
        },
        onSuggestionSelected: (dynamic suggestion) async {
          logger.fine(
              '[cETS - TypeAheadField() -onSuggestionsSelected] - Fired! With suggestion $suggestion');

          await _loadEventFromTemplate(suggestion);

          // Also set our local reactive variables..
          setState(() {
            _templateName = widget.eventData.eventCategory?.templateName;
          });
        });
```

**Comment by KaYBlitZ** [comment]
I tried to copy and paste your code to make it work in the example project but can't replicate the problem

```return Padding(
        padding: EdgeInsets.all(5.0),
        child: Column(
          children: <Widget>[
            Form(
                key: this._formKey,
                child: TypeAheadFormField<dynamic>(
                    getImmediateSuggestions: true,
                    textFieldConfiguration: TextFieldConfiguration<String>(
                      focusNode: _myFocusNode,
                      // autofocus: true,
                      autofocus: false,
                      style: DefaultTextStyle.of(context)
                          .style
                          .copyWith(fontStyle: FontStyle.italic),
                      decoration: InputDecoration(
                        border: const OutlineInputBorder(),
                        filled: true,
                        hintText: 'HINT TEXT',
                      ),
                    ),
                    suggestionsCallback: (String pattern) async {
                      return CitiesService.getSuggestions(pattern);
                    },
                    itemBuilder: (BuildContext context, dynamic suggestion) {
                      // logger.fine('[cETS - TypeAheadField() - itemBuilder] - Fired!');
                      return ListTile(
                        contentPadding: const EdgeInsets.symmetric(
                            vertical: 0.0, horizontal: 15.0),
                        title: Text(
                          suggestion,
                          style: const TextStyle(fontSize: 12.0),
                        ),
                        dense: true,
                        // subtitle: Text(suggestion['descr']),
                      );
                    },
                    validator: (String value) {
                      return 'Please select a template';
                    },
                    onSuggestionSelected: (dynamic suggestion) async {})),
            SizedBox(
              height: 10.0,
            ),
            RaisedButton(
              child: Text('Submit'),
              onPressed: () {
                if (this._formKey.currentState.validate()) {
                  this._formKey.currentState.save();
                  Scaffold.of(context).showSnackBar(SnackBar(
                      content:
                          Text('Your Favorite City is ${this._selectedCity}')));
                }
              },
            ),
          ],
        ));
```

It looks and works well still. 
![screenshot_1551454648](https://user-images.githubusercontent.com/5499214/53648671-7d3eff00-3c0e-11e9-9bf4-d38e4ed20f88.png)

**Comment by KaYBlitZ** [comment]
I am currently working on problems 4/5 with the property to toggle suggestions hide on keyboard hide.

**Comment by sjmcdowall** [comment]
So I wonder what in the world is causing it to NOT work on our project -- as you can see in the image above the suggestion box certainly exceeds past the keyboard .. <scratches head>

**Comment by KaYBlitZ** [comment]
Yea. It is strange. Sorry I can't help. Unable to reproduce.

Offshoot chance, but if you have time, can you try the develop branch from my fork? I moved where findRenderObject method call is actually being called to fix another problem (width resizing issues). It might fix your problem, since that call is used for sizing the suggestions box and I believe it was call prematurely before (hence the width issue).

**Comment by sjmcdowall** [comment]
Yeah I'' give it a whirl since I want to see the other changes too-- but won't get it to until this weekend or later today

**Comment by KaYBlitZ** [comment]
Hey Steve, one of our users @davidmartos96 , found a fix for the problem you are experiencing. The newer versions of Flutter seem to break the the calculation of the keyboard height. That would definitely cause the unscrollable suggestions you are experiencing. He was able to fix the problem by using RootMediaQuery. I was still unable to replicate the problem, but was able to confirm that using RootMediaQuery doesn't break anything. I merged his solution into my develop branch so PR #64 incorporates it. Try it out and feel free to merge it if it fixes your problem and the other additions work.

Thanks again @davidmartos96 .

**Comment by sjmcdowall** [comment]
Will do!! 

On tomorrow’s or this afternoon list of things

**Comment by sjmcdowall** [comment]
Fixed in 1.1.0
